design/one-pager-easy-stacks.md

```diff
@@ -0,0 +1,294 @@
+you need to refer to them in your resource classes that you author for database
+and cluster claims to use.
+
+Today, it's possible to do all this by manually configuring each resource YAML
+files and run `kubectl apply -f <directory of YAMLs>`. However, there are two areas
+that we can improve during this process:
+* The preparation phase of these YAMLs requires to be very careful and changing
+  something usually means touching a few different files. While this may not seem
+  like a big hurdle, changing some high level parameters of that environment is
+  cumbersome _after_ it's deployed.
+* A user who is familiar with AWS may not be familiar with Azure or GCP; so,
+  ready-made configuration stacks that promise the exact same setup would be
+  useful for them to see how they can achieve similar setups and run various
+  benchmarks according to their business needs.
+* After you run `kubectl apply -f <directory of YAMLs>`, a dependency graph of
+  resources starts to resolve. However, it's not always clear for user to know _when_
+  everything is actually ready to use unless they know readiness of which resources
+  signal the readiness of the whole environment.
+
+## Goals
+
+The main goal of this design is to make it easier for people to write sets of
+configurations and be able to manage those configuration sets on a higher level.
+
+It is important that the design puts forward:
+
+* A base boilerplate package where creating a new configuration set that has a
+  controller is as easy as changing the YAML files.
+* A way to declare pre-requisites that needs to be satisfied by the cluster
+  for the configuration set to work.
+* A way to change a high level configuration by changing 1 value that is then
+  translated to the YAMLs that are already deployed.
+
+## Proposal
+
+Crossplane has the notion of stacks that can be used as extension to Crossplane's
+abilites, to deploy an application and all other things that you can do with
+Kubernetes controllers since main components of stacks are CRDs and controllers
+watching them.
+
+To achieve the mentioned goals, we can publish the YAML files in a stack with a
+controller that has its own CRD with high level parameters. Controller deploys
+and updates the YAMLs that are packaged in the stack image according to user's
+input on the CR instance of that stack that user creates.
+
+As sky is the limit when it comes to stacks, we need to put down some general
+rules to be adhered by initial configuration stacks.
+
+* CRDs of the resources that will be deployed by the stack are declared as
+  dependency so that user knows what CR instances will be deployed by the given
+  configuration stack.
+* It should be possible to have different sets of configurations in one stack
+  and user should be able to choose from them via the CR that is an instance
+  of stack's CRD.
+  * Note that it's highly preferred to have 1 custom resource definition that
```

Comment:
> Note that it's highly preferred to have 1 custom resource definition

What is the reason for this preference?

Reply:
Also, I am thinking that we may want to require that it is a single CRD at this time, because it seems like in order to support multiple we would need a lot of the configuration that we are designing as part of the template stacks work.

Reply:
This has to do with how we name the deployed resources. Basically, when you have only one CRD, you can just prefix the resources with the CR instance name. The api-server would reject creation of CR instances with the same name, so, name as prefix is enough. When you have multiple CRDs, name is not unique anymore. So, you either use `uid` or some other method for making deployed resources exclusive to one CR that intended to deploy them.

However, it's also possible that you can have multiple CRDs that deploy completely different kind sets of resources, you'd not need to worry about the name collision in that case. So, it's not that you should never do it but it's just highly recommended to go with one custom resource definition and build another stack for that extra CRD that you want.